src/commands/buffer.rs

```diff
@@ -468,6 +469,127 @@ pub fn outdent_line(app: &mut Application) -> Result {
     Ok(())
 }
 
+pub fn toggle_line_comment(app: &mut Application) -> Result {
+    lazy_static! {
+        // The `nerdcommenter` plugin for vim has a nice, comprehensive collection
+        // of possible languages and their commment syntaxes.
+        // https://github.com/preservim/nerdcommenter/blob/253eafd3a7/plugin/NERD_commenter.vim#L58
+        //
+        // If you want to add a new entry and don't know the correct scope name,
+        // you can use the command `buffer::display_current_scope`.
+        static ref COMMENT_MAP: HashMap<Scope, &'static str> = {
+            let mut m = HashMap::new();
+            m.insert(Scope::new("embedding.php").unwrap(), "// ");
+            m.insert(Scope::new("source.c").unwrap(), "// ");
+            m.insert(Scope::new("source.c++").unwrap(), "// ");
+            m.insert(Scope::new("source.cmake").unwrap(), "# ");
+            m.insert(Scope::new("source.go").unwrap(), "// ");
+            m.insert(Scope::new("source.js").unwrap(), "// ");
+            m.insert(Scope::new("source.lua").unwrap(), "-- ");
+            m.insert(Scope::new("source.makefile").unwrap(), "# ");
+            m.insert(Scope::new("source.python").unwrap(), "# ");
+            m.insert(Scope::new("source.rust").unwrap(), "// ");
+            m.insert(Scope::new("source.shell.bash").unwrap(), "# ");
+            m.insert(Scope::new("source.swift").unwrap(), "// ");
+            m.insert(Scope::new("source.toml").unwrap(), "# ");
+            m.insert(Scope::new("source.yaml").unwrap(), "# ");
+            m.insert(Scope::new("text.tex").unwrap(), "% ");
+            m.insert(Scope::new("text.tex.latex").unwrap(), "% ");
+            m
+        };
+    }
+
+    let buffer = app.workspace.current_buffer().ok_or(BUFFER_MISSING)?;
+    let original_cursor = *buffer.cursor.clone();
+
+    let comment_content = {
+        let scope = buffer.syntax_definition.as_ref()
+            .map_or(Scope::new("none").unwrap(), |def| def.scope);
+
+        COMMENT_MAP.get(&scope).ok_or_else(|| {
+            format!("No comment syntax definition found for scope {}", scope)
+        })?
+    };
+
+    // Get the range of lines we'll comment based on
+    // either the current selection or cursor line.
+    let lines = match app.mode {
+        Mode::SelectLine(ref mode) => {
+            if mode.anchor >= buffer.cursor.line {
+                buffer.cursor.line..mode.anchor + 1
+            } else {
+                mode.anchor..buffer.cursor.line + 1
+            }
+        }
+        _ => buffer.cursor.line..buffer.cursor.line + 1,
+    };
+
+    let buffer_range = Range::new(
+        Position { line: lines.start, offset: 0 },
+        Position { line: lines.end, offset: 0 }
+    );
+
+    let buffer_range_content = buffer.read(&buffer_range).ok_or(CURRENT_LINE_MISSING)?;
+
+    // Split the buffer content on newlines and filter out any zero-length lines.
```

Comment:
While we _are_ using this logic to transform the line data portion of this, the `zip` is subtle and it's not entirely clear what we're producing at the end of this operation.

What do you think of something along the lines of "produce a collection of line numbers and content, but only for non-empty lines"?

Reply:
Sounds good, going to point out that operation explicitly.